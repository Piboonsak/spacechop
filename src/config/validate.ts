import Config from '../types/Config';

export const validate = (config) => {
  try {
    Config.check(config);
  } catch (err) {
    throw new Error(`${err.key}: ${err.message}`);
  }
};

let valid = true;

export default (config) => {
  try {
    validate(config);
    if (!valid) {
      valid = true;
      console.log('Config is valid again');
    }
    return true;
  } catch (err) {
    valid = false;
    console.log('\nThere is an error in /config.yml:');
    console.log('--------------------------------');
    console.log(err.message);
    console.log('--------------------------------');
  }
  return false;
};
