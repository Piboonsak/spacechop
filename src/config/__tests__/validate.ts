import { validate } from '../validate';

describe('should validate config.yml', () => {
  it('should require paths', () => {
    const config = {};
    expect(() => validate(config)).toThrowErrorMatchingSnapshot();
  });

  it('should require valid paths array', () => {
    let config = { paths: {}, sources: {}, presets: {} };
    expect(() => validate(config)).toThrowErrorMatchingSnapshot();
    config = { paths: [], sources: {}, presets: {} };
    expect(() => validate(config)).toThrowErrorMatchingSnapshot();
    config = { paths: [''], sources: {}, presets: {} };
    expect(() => validate(config)).toThrowErrorMatchingSnapshot();
    config = { paths: ['/'], sources: {}, presets: {} };
    expect(() => validate(config)).toThrowErrorMatchingSnapshot();
    config = { paths: ['/:image'], sources: {}, presets: {} };
    expect(() => validate(config)).toThrowErrorMatchingSnapshot();
    config = { paths: [':preset'], sources: {}, presets: {} };
    expect(() => validate(config)).toThrowErrorMatchingSnapshot();
    config = {
      paths: ['/:preset'],
      sources: [{ http: { root: 'http://localhost' } }],
      presets: {
        t_200: {
          steps: [{
            $crop: {
              width: 200,
              height: 200,
            },
          }],
        },
      },
    };
    expect(() => validate(config)).not.toThrow();
    config = {
      paths: ['/:preset/:image'],
      sources: [{ http: { root: 'http://localhost' } }],
      presets: {
        t_200: {
          steps: [{
            $crop: {
              width: 200,
              height: 200,
            },
          }],
        },
      },
    };
    expect(() => validate(config)).not.toThrow();
  });

  it('should require sources', () => {
    const config = { paths: ['/:preset/:image'] };
    expect(() => validate(config)).toThrowErrorMatchingSnapshot();
  });

  it('should require valid sources array', () => {
    const config = { paths: ['/:preset/:image'], sources: {} };
    expect(() => validate(config)).toThrowErrorMatchingSnapshot();
  });

  it('should require valid source', () => {
    const config = { paths: ['/:preset/:image'], sources: [{}] };
    expect(() => validate(config)).toThrowErrorMatchingSnapshot();
  });

  it('should require valid source http', () => {
    const config = { paths: ['/:preset/:image'], sources: [{ http: {} }] };
    expect(() => validate(config)).toThrowErrorMatchingSnapshot();
  });

  it('should require valid http root url', () => {
    const config = { paths: ['/:preset/:image'], sources: [{ http: { root: '' } }] };
    expect(() => validate(config)).toThrowErrorMatchingSnapshot();
  });

  it('should require valid source s3', () => {
    let config = { paths: ['/:preset/:image'], sources: [{ s3: {} }], presets: {} };
    expect(() => validate(config)).toThrowErrorMatchingSnapshot();
    config = { paths: ['/:preset/:image'], sources: [{ s3: {
      access_key_id: '',
      secret_access_key: '',
      region: '',
      bucket_name: '',
      path: '',
    } }], presets: {} };
    expect(() => validate(config)).toThrowErrorMatchingSnapshot();
    config = {
      paths: ['/:preset/:image'],
      sources: [{ s3: {
        access_key_id: 'access_key_id',
        secret_access_key: 'secret_access_key',
        region: 'region',
        bucket_name: 'bucket_name',
        path: '',
      } }],
      presets: {
        t_200: {
          steps: [{
            $crop: {
              width: 200,
              height: 200,
            },
          }],
        },
      },
    };
    expect(() => validate(config)).not.toThrow();
  });

  it('should throw on any other type of source', () => {
    const config = { paths: ['/:preset/:image'], sources: [{ bad_source: {} }] };
    expect(() => validate(config)).toThrowErrorMatchingSnapshot();
  });

  it('should require presets', () => {
    const config = { paths: ['/:preset/:image'], sources: [{ http: { root: 'http://localhost' } }] };
    expect(() => validate(config)).toThrowErrorMatchingSnapshot();
  });

  it('should require valid presets', () => {
    let config = {
      paths: ['/:preset/:image'],
      sources: [{ http: { root: 'http://localhost' } }],
      presets: {
        t_200: {},
      },
    };
    expect(() => validate(config)).toThrowErrorMatchingSnapshot();
    config = {
      paths: ['/:preset/:image'],
      sources: [{ http: { root: 'http://localhost' } }],
      presets: {
        t_200: {
          steps: {},
        },
      },
    };
    expect(() => validate(config)).toThrowErrorMatchingSnapshot();
    config = {
      paths: ['/:preset/:image'],
      sources: [{ http: { root: 'http://localhost' } }],
      presets: {
        t_200: {
          steps: [{}],
        },
      },
    };
    expect(() => validate(config)).toThrowErrorMatchingSnapshot();
    config = {
      paths: ['/:preset/:image'],
      sources: [{ http: { root: 'http://localhost' } }],
      presets: {
        t_200: {
          steps: [{
            $crop: {},
          }],
        },
      },
    };
    expect(() => validate(config)).toThrowErrorMatchingSnapshot();
    config = {
      paths: ['/:preset/:image'],
      sources: [{ http: { root: 'http://localhost' } }],
      presets: {
        t_200: {
          steps: [{
            $crop: {
              width: 200,
              height: 200,
            },
          }],
        },
      },
    };
    expect(() => validate(config)).not.toThrow();
    config = {
      paths: ['/:preset/:image'],
      sources: [{ http: { root: 'http://localhost' } }],
      presets: {
        t_200: {
          steps: [{
            $crop: {
              width: 200,
              height: 200,
              gravity: 'center',
            },
          }],
        },
      },
    };
    expect(() => validate(config)).not.toThrow();
    config = {
      paths: ['/:preset/:image'],
      sources: [{ http: { root: 'http://localhost' } }],
      presets: {
        t_200: {
          steps: [{
            $crop: {
              width: 200,
              height: 200,
              gravity: 'left',
            },
          }],
        },
      },
    };
    expect(() => validate(config)).toThrowErrorMatchingSnapshot();
    config = {
      paths: ['/:preset/:image'],
      sources: [{ http: { root: 'http://localhost' } }],
      presets: {
        t_200: {
          steps: [{
            $crop: {
              width: 200,
              height: 200,
              gravity: 'face',
            },
          }],
        },
      },
    };
    expect(() => validate(config)).not.toThrow();
    config = {
      paths: ['/:preset/:image'],
      sources: [{ http: { root: 'http://localhost' } }],
      presets: {
        t_200: {
          steps: [{
            $compress: {},
          }],
        },
      },
    };
    expect(() => validate(config)).not.toThrow();
    config = {
      paths: ['/:preset/:image'],
      sources: [{ http: { root: 'http://localhost' } }],
      presets: {
        t_200: {
          steps: [{
            $compress: {
              lossy: true,
            },
          }],
        },
      },
    };
    expect(() => validate(config)).not.toThrow();
    config = {
      paths: ['/:preset/:image'],
      sources: [{ http: { root: 'http://localhost' } }],
      presets: {
        t_200: {
          steps: [{
            $compress: {
              quality: 0.9,
            },
          }],
        },
      },
    };
    expect(() => validate(config)).not.toThrow();
    config = {
      paths: ['/:preset/:image'],
      sources: [{ http: { root: 'http://localhost' } }],
      presets: {
        t_200: {
          steps: [{
            $compress: {
              strip: true,
            },
          }],
        },
      },
    };
    expect(() => validate(config)).not.toThrow();
    config = {
      paths: ['/:preset/:image'],
      sources: [{ http: { root: 'http://localhost' } }],
      presets: {
        t_200: {
          steps: [{
            $compress: {
              quality: 0.9,
              strip: true,
              lossy: true,
            },
          }],
        },
      },
    };
    expect(() => validate(config)).not.toThrow();
    config = {
      paths: ['/:preset/:image'],
      sources: [{ http: { root: 'http://localhost' } }],
      presets: {
        t_200: {
          steps: [{
            $fill: {
              width: 200,
              height: 200,
              gravity: 'center',
            },
          }],
        },
      },
    };
    expect(() => validate(config)).not.toThrow();
    config = {
      paths: ['/:preset/:image'],
      sources: [{ http: { root: 'http://localhost' } }],
      presets: {
        t_200: {
          steps: [{
            $fit: {
              width: 200,
              height: 200,
              gravity: 'northeast',
            },
          }],
        },
      },
    };
    expect(() => validate(config)).not.toThrow();
    config = {
      paths: ['/:preset/:image'],
      sources: [{ http: { root: 'http://localhost' } }],
      presets: {
        t_200: {
          steps: [{
            $resize: {
              width: 200,
              height: 200,
            },
          }],
        },
      },
    };
    expect(() => validate(config)).not.toThrow();
  });
});
