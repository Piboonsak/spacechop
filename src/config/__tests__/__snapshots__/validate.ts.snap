// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should validate config.yml should require paths 1`] = `"paths: Expected array, but was undefined"`;

exports[`should validate config.yml should require presets 1`] = `"presets: Expected { [_: string]: { steps: ({ $compress: { quality?: number; lossy?: boolean; }; } | { $crop: { width?: number; height?: number; gravity?: \\"center\\" | \\"face\\" | \\"north\\" | \\"northeast\\" | \\"east\\" | \\"southeast\\" | \\"south\\" | \\"southwest\\" | \\"west\\" | \\"northwest\\"; }; } | { $fill: { width: number; height: number; gravity: undefined | (\\"center\\" | \\"face\\" | \\"north\\" | \\"northeast\\" | \\"east\\" | \\"southeast\\" | \\"south\\" | \\"southwest\\" | \\"west\\" | \\"northwest\\"); }; } | { $fit: { width?: number; height?: number; gravity?: \\"center\\" | \\"face\\" | \\"north\\" | \\"northeast\\" | \\"east\\" | \\"southeast\\" | \\"south\\" | \\"southwest\\" | \\"west\\" | \\"northwest\\"; }; } | { $format: { type?: \\"jpeg\\" | \\"png\\" | \\"gif\\" | \\"webp\\"; }; } | { $resize: { width: number; height: number; }; } | { $strip: { icu_profile?: boolean; }; })[]; } }, but was undefined"`;

exports[`should validate config.yml should require sources 1`] = `"sources: Expected array, but was undefined"`;

exports[`should validate config.yml should require valid http root url 1`] = `"sources.[0]: Expected { http: { root: string; }; } | { s3: { access_key_id: string; secret_access_key: string; region: string; bucket_name: string; path: string; }; }, but was object"`;

exports[`should validate config.yml should require valid paths array 1`] = `"paths: Expected array, but was object"`;

exports[`should validate config.yml should require valid paths array 2`] = `"paths: Requires at least one path"`;

exports[`should validate config.yml should require valid paths array 3`] = `"paths.[0]: Cannot be empty"`;

exports[`should validate config.yml should require valid paths array 4`] = `"paths.[0]: Requires :preset in path"`;

exports[`should validate config.yml should require valid paths array 5`] = `"paths.[0]: Requires :preset in path"`;

exports[`should validate config.yml should require valid paths array 6`] = `"paths.[0]: Must start with /"`;

exports[`should validate config.yml should require valid presets 1`] = `"presets.t_200.steps: Expected array, but was undefined"`;

exports[`should validate config.yml should require valid presets 2`] = `"presets.t_200.steps: Expected array, but was object"`;

exports[`should validate config.yml should require valid presets 3`] = `"presets.t_200.steps.[0]: Expected { $compress: { quality?: number; lossy?: boolean; }; } | { $crop: { width?: number; height?: number; gravity?: \\"center\\" | \\"face\\" | \\"north\\" | \\"northeast\\" | \\"east\\" | \\"southeast\\" | \\"south\\" | \\"southwest\\" | \\"west\\" | \\"northwest\\"; }; } | { $fill: { width: number; height: number; gravity: undefined | (\\"center\\" | \\"face\\" | \\"north\\" | \\"northeast\\" | \\"east\\" | \\"southeast\\" | \\"south\\" | \\"southwest\\" | \\"west\\" | \\"northwest\\"); }; } | { $fit: { width?: number; height?: number; gravity?: \\"center\\" | \\"face\\" | \\"north\\" | \\"northeast\\" | \\"east\\" | \\"southeast\\" | \\"south\\" | \\"southwest\\" | \\"west\\" | \\"northwest\\"; }; } | { $format: { type?: \\"jpeg\\" | \\"png\\" | \\"gif\\" | \\"webp\\"; }; } | { $resize: { width: number; height: number; }; } | { $strip: { icu_profile?: boolean; }; }, but was object"`;

exports[`should validate config.yml should require valid presets 4`] = `"presets.t_200.steps.[0]: Expected { $compress: { quality?: number; lossy?: boolean; }; } | { $crop: { width?: number; height?: number; gravity?: \\"center\\" | \\"face\\" | \\"north\\" | \\"northeast\\" | \\"east\\" | \\"southeast\\" | \\"south\\" | \\"southwest\\" | \\"west\\" | \\"northwest\\"; }; } | { $fill: { width: number; height: number; gravity: undefined | (\\"center\\" | \\"face\\" | \\"north\\" | \\"northeast\\" | \\"east\\" | \\"southeast\\" | \\"south\\" | \\"southwest\\" | \\"west\\" | \\"northwest\\"); }; } | { $fit: { width?: number; height?: number; gravity?: \\"center\\" | \\"face\\" | \\"north\\" | \\"northeast\\" | \\"east\\" | \\"southeast\\" | \\"south\\" | \\"southwest\\" | \\"west\\" | \\"northwest\\"; }; } | { $format: { type?: \\"jpeg\\" | \\"png\\" | \\"gif\\" | \\"webp\\"; }; } | { $resize: { width: number; height: number; }; } | { $strip: { icu_profile?: boolean; }; }, but was object"`;

exports[`should validate config.yml should require valid presets 5`] = `"presets.t_200.steps.[0]: Expected { $compress: { quality?: number; lossy?: boolean; }; } | { $crop: { width?: number; height?: number; gravity?: \\"center\\" | \\"face\\" | \\"north\\" | \\"northeast\\" | \\"east\\" | \\"southeast\\" | \\"south\\" | \\"southwest\\" | \\"west\\" | \\"northwest\\"; }; } | { $fill: { width: number; height: number; gravity: undefined | (\\"center\\" | \\"face\\" | \\"north\\" | \\"northeast\\" | \\"east\\" | \\"southeast\\" | \\"south\\" | \\"southwest\\" | \\"west\\" | \\"northwest\\"); }; } | { $fit: { width?: number; height?: number; gravity?: \\"center\\" | \\"face\\" | \\"north\\" | \\"northeast\\" | \\"east\\" | \\"southeast\\" | \\"south\\" | \\"southwest\\" | \\"west\\" | \\"northwest\\"; }; } | { $format: { type?: \\"jpeg\\" | \\"png\\" | \\"gif\\" | \\"webp\\"; }; } | { $resize: { width: number; height: number; }; } | { $strip: { icu_profile?: boolean; }; }, but was object"`;

exports[`should validate config.yml should require valid source 1`] = `"sources.[0]: Expected { http: { root: string; }; } | { s3: { access_key_id: string; secret_access_key: string; region: string; bucket_name: string; path: string; }; }, but was object"`;

exports[`should validate config.yml should require valid source http 1`] = `"sources.[0]: Expected { http: { root: string; }; } | { s3: { access_key_id: string; secret_access_key: string; region: string; bucket_name: string; path: string; }; }, but was object"`;

exports[`should validate config.yml should require valid source s3 1`] = `"sources.[0]: Expected { http: { root: string; }; } | { s3: { access_key_id: string; secret_access_key: string; region: string; bucket_name: string; path: string; }; }, but was object"`;

exports[`should validate config.yml should require valid source s3 2`] = `"sources.[0]: Expected { http: { root: string; }; } | { s3: { access_key_id: string; secret_access_key: string; region: string; bucket_name: string; path: string; }; }, but was object"`;

exports[`should validate config.yml should require valid sources array 1`] = `"sources: Expected array, but was object"`;

exports[`should validate config.yml should throw on any other type of source 1`] = `"sources.[0]: Expected { http: { root: string; }; } | { s3: { access_key_id: string; secret_access_key: string; region: string; bucket_name: string; path: string; }; }, but was object"`;
